File: sdk/javascript/src/validate.js
/**
 * Schema validation for FoodBlocks (Section 8).
 *
 * Validates a block's state against a schema definition.
 * Validation is always optional — a block without a $schema reference is valid.
 */

// Bundled core schemas (subset — full registry is fetched from the network)
const CORE_SCHEMAS = {
  'foodblock:substance.product@1.0': {
    target_type: 'substance.product',
    version: '1.0',
    fields: {
      name: { type: 'string', required: true },
      price: { type: 'number' },
      unit: { type: 'string' },
      weight: { type: 'object' },
      allergens: { type: 'object' },
      gtin: { type: 'string' }
    },
    expected_refs: ['seller'],
    optional_refs: ['origin', 'inputs', 'certifications'],
    requires_instance_id: false
  },
  'foodblock:transfer.order@1.0': {
    target_type: 'transfer.order',
    version: '1.0',
    fields: {
      instance_id: { type: 'string', required: true },
      quantity: { type: 'number' },
      unit: { type: 'string' },
      total: { type: 'number' },
      payment_ref: { type: 'string' }
    },
    expected_refs: ['buyer', 'seller'],
    optional_refs: ['product', 'agent'],
    requires_instance_id: true
  },
  'foodblock:observe.review@1.0': {
    target_type: 'observe.review',
    version: '1.0',
    fields: {
      instance_id: { type: 'string', required: true },
      rating: { type: 'number', required: true },
      text: { type: 'string' }
    },
    expected_refs: ['subject', 'author'],
    optional_refs: [],
    requires_instance_id: true
  },
  'foodblock:actor.producer@1.0': {
    target_type: 'actor.producer',
    version: '1.0',
    fields: {
      name: { type: 'string', required: true },
      public_key_sign: { type: 'string' },
      public_key_encrypt: { type: 'string' },
      gln: { type: 'string' }
    },
    expected_refs: [],
    optional_refs: [],
    requires_instance_id: false
  },
  'foodblock:observe.certification@1.0': {
    target_type: 'observe.certification',
    version: '1.0',
    fields: {
      instance_id: { type: 'string', required: true },
      name: { type: 'string', required: true },
      valid_until: { type: 'string' },
      standard: { type: 'string' }
    },
    expected_refs: ['subject', 'authority'],
    optional_refs: [],
    requires_instance_id: true
  }
}

/**
 * Validate a block against its declared schema or a provided schema.
 *
 * @param {object} block - The FoodBlock to validate
 * @param {object} [schema] - Optional schema object. If not provided, uses block's $schema field.
 * @param {object} [registry] - Optional custom schema registry
 * @returns {string[]} - Array of error messages (empty = valid)
 */
function validate(block, schema, registry) {
  const errors = []

  if (!block || !block.type || !block.state) {
    errors.push('Block must have type and state')
    return errors
  }

  // Resolve schema
  let schemaDef = schema
  if (!schemaDef && block.state.$schema) {
    const reg = registry || CORE_SCHEMAS
    schemaDef = reg[block.state.$schema]
    if (!schemaDef) {
      errors.push(`Unknown schema: ${block.state.$schema}`)
      return errors
    }
  }

  // No schema to validate against — block is valid
  if (!schemaDef) {
    return errors
  }

  // Check type match
  if (schemaDef.target_type && block.type !== schemaDef.target_type) {
    errors.push(`Type mismatch: block is ${block.type}, schema is for ${schemaDef.target_type}`)
  }

  // Check required fields
  if (schemaDef.fields) {
    for (const [field, def] of Object.entries(schemaDef.fields)) {
      if (def.required && block.state[field] === undefined) {
        errors.push(`Missing required field: state.${field}`)
      }
      if (block.state[field] !== undefined && def.type) {
        const actualType = typeof block.state[field]
        if (def.type === 'object' && (actualType !== 'object' || Array.isArray(block.state[field]))) {
          errors.push(`Field state.${field} should be ${def.type}, got ${actualType}`)
        } else if (def.type !== 'object' && actualType !== def.type) {
          errors.push(`Field state.${field} should be ${def.type}, got ${actualType}`)
        }
      }
    }
  }

  // Check required refs
  if (schemaDef.expected_refs) {
    for (const ref of schemaDef.expected_refs) {
      if (!block.refs || block.refs[ref] === undefined) {
        errors.push(`Missing expected ref: refs.${ref}`)
      }
    }
  }

  // Check instance_id requirement
  if (schemaDef.requires_instance_id && !block.state.instance_id) {
    errors.push('Missing required field: state.instance_id')
  }

  return errors
}

module.exports = { validate, CORE_SCHEMAS }

File: sdk/go/validate.go
package foodblock

import "fmt"

// SchemaField defines a field in a schema.
type SchemaField struct {
	Type     string
	Required bool
}

// Schema defines validation rules for a block type.
type Schema struct {
	TargetType        string
	Version           string
	Fields            map[string]SchemaField
	ExpectedRefs      []string
	OptionalRefs      []string
	RequiresInstanceID bool
}

// CoreSchemas are the bundled core schemas.
var CoreSchemas = map[string]Schema{
	"foodblock:substance.product@1.0": {
		TargetType: "substance.product",
		Version:    "1.0",
		Fields: map[string]SchemaField{
			"name":  {Type: "string", Required: true},
			"price": {Type: "number"},
			"unit":  {Type: "string"},
		},
		ExpectedRefs:      []string{"seller"},
		OptionalRefs:      []string{"origin", "inputs", "certifications"},
		RequiresInstanceID: false,
	},
	"foodblock:transfer.order@1.0": {
		TargetType: "transfer.order",
		Version:    "1.0",
		Fields: map[string]SchemaField{
			"instance_id": {Type: "string", Required: true},
			"quantity":    {Type: "number"},
			"unit":        {Type: "string"},
			"total":       {Type: "number"},
		},
		ExpectedRefs:      []string{"buyer", "seller"},
		OptionalRefs:      []string{"product", "agent"},
		RequiresInstanceID: true,
	},
	"foodblock:observe.review@1.0": {
		TargetType: "observe.review",
		Version:    "1.0",
		Fields: map[string]SchemaField{
			"instance_id": {Type: "string", Required: true},
			"rating":      {Type: "number", Required: true},
			"text":        {Type: "string"},
		},
		ExpectedRefs:      []string{"subject", "author"},
		RequiresInstanceID: true,
	},
	"foodblock:actor.producer@1.0": {
		TargetType: "actor.producer",
		Version:    "1.0",
		Fields: map[string]SchemaField{
			"name": {Type: "string", Required: true},
		},
		RequiresInstanceID: false,
	},
	"foodblock:observe.certification@1.0": {
		TargetType: "observe.certification",
		Version:    "1.0",
		Fields: map[string]SchemaField{
			"instance_id": {Type: "string", Required: true},
			"name":        {Type: "string", Required: true},
			"valid_until": {Type: "string"},
			"standard":    {Type: "string"},
		},
		ExpectedRefs:      []string{"subject", "authority"},
		RequiresInstanceID: true,
	},
}

// Validate validates a block against a schema. Returns a list of error messages (empty = valid).
func Validate(block Block, schema *Schema) []string {
	var errs []string

	if block.Type == "" {
		errs = append(errs, "Block must have type and state")
		return errs
	}

	// Resolve schema from block's $schema field if not provided
	schemaDef := schema
	if schemaDef == nil {
		if schemaRef, ok := block.State["$schema"].(string); ok {
			if s, exists := CoreSchemas[schemaRef]; exists {
				schemaDef = &s
			} else {
				errs = append(errs, fmt.Sprintf("Unknown schema: %s", schemaRef))
				return errs
			}
		}
	}

	if schemaDef == nil {
		return errs
	}

	// Check type match
	if schemaDef.TargetType != "" && block.Type != schemaDef.TargetType {
		errs = append(errs, fmt.Sprintf("Type mismatch: block is %s, schema is for %s", block.Type, schemaDef.TargetType))
	}

	// Check required fields
	for field, def := range schemaDef.Fields {
		if def.Required {
			if _, ok := block.State[field]; !ok {
				errs = append(errs, fmt.Sprintf("Missing required field: state.%s", field))
			}
		}
		if val, ok := block.State[field]; ok && def.Type != "" {
			actualType := goTypeToSchemaType(val)
			if actualType != def.Type {
				errs = append(errs, fmt.Sprintf("Field state.%s should be %s, got %s", field, def.Type, actualType))
			}
		}
	}

	// Check required refs
	for _, ref := range schemaDef.ExpectedRefs {
		if _, ok := block.Refs[ref]; !ok {
			errs = append(errs, fmt.Sprintf("Missing expected ref: refs.%s", ref))
		}
	}

	// Check instance_id requirement
	if schemaDef.RequiresInstanceID {
		if _, ok := block.State["instance_id"]; !ok {
			errs = append(errs, "Missing required field: state.instance_id")
		}
	}

	return errs
}

func goTypeToSchemaType(v interface{}) string {
	switch v.(type) {
	case string:
		return "string"
	case float64, int, int64:
		return "number"
	case bool:
		return "boolean"
	case map[string]interface{}:
		return "object"
	case []interface{}:
		return "array"
	default:
		return "unknown"
	}
}
